### PR TITLE
AOS-51: Distinguish Errors by IsAosError

### PR DESCRIPTION
**Pull request by xxchan**

### CODE REVIEW

File: rfcs/51-identify-go-storage-error.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+## Rationale
+
+We cannot add methods to sentinel errors, so we need to replace them.
+
+### Alternative
+
+Add an interface and use type cast to judge.
+
+```go
+type ErrGoStorage interface {
```
Comment (Xuanwo) [inline rfcs/51-identify-go-storage-error.md:52]: How define an internal interface like following:

```go
type AosError interface {
	isAosError()
}
```

We will have code like the following:

```go
type Error string

func (e Error) Error() string {
	return string(e)
}

func (e Error) isAosError() {}

type AosError interface {
	isAosError()
}
```

We only need to add an empty function for our errors.

Here are benchmark tests:

```go
import (
	"errors"
	"testing"
)

type Error string

var specialError = errors.New("x")

func (e Error) Error() string {
	return string(e)
}

func (e Error) Is(err error) bool {
	return true
}

func (e Error) isAosError() {}

type AosError interface {
	isAosError()
}

func BenchmarkInternalInterface(b *testing.B) {
	var e error
	e = Error("xxxx")

	for i := 0; i < b.N; i++ {
		_ = e.(AosError)
	}
}

func BenchmarkErrorsIs(b *testing.B) {
	var e error
	e = Error("xxxx")

	for i := 0; i < b.N; i++ {
		_ = errors.Is(e, specialError)
	}
}

Resultes:

```
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkInternalInterface
BenchmarkInternalInterface-8   	176883178	         6.845 ns/op
BenchmarkErrorsIs
BenchmarkErrorsIs-8            	78588775	        13.68 ns/op
PASS
```

Use our own interface is a bit faster.

Comment (xxchan) [reply rfcs/51-identify-go-storage-error.md:52]: Benchmark is convincing! I will also try to use benchmark to compare alternatives in the future :p

Comment (xxchan) [reply rfcs/51-identify-go-storage-error.md:52]: If `isAosError` is an unexported method, what if a service creates its error `struct`? Then it cannot implement `AosError`

Comment (Xuanwo) [reply rfcs/51-identify-go-storage-error.md:52]: They should define new error by using `services.Error` which already implement it.

Comment (xxchan) [reply rfcs/51-identify-go-storage-error.md:52]: A workaround is to embed `ErrorCode` as an anonymous field into error `struct`s... (The hidden `Err` approach)

Btw, I was just considering making `ErrorCode` private... Just like `errors.New()` returns private `stringError` 

So two choices:
1. private `isAosError` + public `ErrorCode` + embed `ErrorCode` in `struct`
2. public `IsAosError` + private (or public) `errorCode` + no embed

Comment (Xuanwo) [reply rfcs/51-identify-go-storage-error.md:52]: > If `isAosError` is an unexported method, what if a service creates its error `struct`? Then it cannot implement `AosError`.

It OK to implement an interface even there is an unexported method. I make this function private is to restrict user's ability.

Comment (Xuanwo) [reply rfcs/51-identify-go-storage-error.md:52]: Let's move the `AosError` definition to another package:

```go
func (e Error) Error() string {
	return string(e)
}

func (e Error) Is(err error) bool {
	return true
}

func (e Error) isAosError() {}

func BenchmarkInternalInterface(b *testing.B) {
	var e error
	e = Error("xxxx")

	for i := 0; i < b.N; i++ {
		_, _ = e.(types.AosError)
	}
}
```

It still works.

Comment (xxchan) [reply rfcs/51-identify-go-storage-error.md:52]: Take a look at `_, ok := e.(a.AosError)`, `ok` is `false`

Comment (Xuanwo) [reply rfcs/51-identify-go-storage-error.md:52]: Oh, I made a mistake.

Comment (xxchan) [reply rfcs/51-identify-go-storage-error.md:52]:
```go
var e (a.AosError)
e = Error("xxxx")
```
can find problem more quickly

Comment (Xuanwo) [reply rfcs/51-identify-go-storage-error.md:52]: OK, let's export `IsAosError`, but with comment like:

```go
// IsAosError should only be implemented by go-storage & go-service.
// We depends this function and related AosError interface to detect error.
```

**Comment by xxchan** [comment]
I think there's no need to implement interface for top-level `ServiceError` & `StorageError` & `InitError`. What's your opinion? @Xuanwo

**Comment by Xuanwo** [comment]
> I think there's no need to implement interface for top-level `ServiceError` & `StorageError` & `InitError`. What's your opinion? @Xuanwo

But `newStorager` will return an `InitError` which could be handled by `formatError`.

**Comment by xxchan** [comment]
> But `newStorager` will return an `InitError` which could be handled by `formatError`.

I think (and checked) this is not true:

`New*` just call `new*`, and they are both in `utils.go`. `new*` have a `defer` where set `err = &InitError{...}`

```go
func newServicer(pairs ...typ.Pair) (srv *Service, err error) {
	defer func() {
		if err != nil {
			err = &services.InitError{Op: "new_servicer", Type: Type, Err: err, Pairs: pairs}
		}
	}()
```
Wait, I think it should be `Err: formatError(err)`

**Comment by xxchan** [comment]
BTW, I found I missed `credential.Error`, `endpoint.Error`, etc, in AOS-47.

**Comment by Xuanwo** [comment]
> > But `newStorager` will return an `InitError` which could be handled by `formatError`.
> 
> I think (and checked) this is not true:
> 
> `New*` just call `new*`, and they are both in `utils.go` (hand-written). `new*` have a `defer` where set `err = &InitError{...}`
> 
> ```go
> func newServicer(pairs ...typ.Pair) (srv *Service, err error) {
> 	defer func() {
> 		if err != nil {
> 			err = &services.InitError{Op: "new_servicer", Type: Type, Err: err, Pairs: pairs}
> 		}
> 	}()
> ```


Oh, nice catch, so `InitError` is returned directly without wrapped into `ServiceError` or `StorageError`.

> Wait, I think it should be `Err: formatError(err)`

No, I think the err in `InitError` should not be formatted, as those errors are not SDK?

**Comment by xxchan** [comment]
> No, I think the err in `InitError` should not be formatted, as those errors are not SDK?

Most not, but still exist:

```go

func newServicer(pairs ...typ.Pair) (srv *Service, err error) {
	// ...
	
	switch cp.Protocol() {
	case credential.ProtocolHmac:
		cfg, err = qsconfig.New(cp.Hmac())  // HERE!
		if err != nil {
			return nil, err
		}
		// ...
	}
	//...

	// func service.Init(c *qsconfig.Config) (*service.Service, error)
	// why err is omited here?
	srv.service, _ = service.Init(cfg)  

	// ...
}

```

**Comment by Xuanwo** [comment]
OK, those errors that after `formatError` will be `UnexpectedError`, LGTM.

File: rfcs/51-identify-go-storage-error.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-05-10
+---
+
+# AOS-51: Identify go-storage Error
```
Comment (Xuanwo) [inline rfcs/51-identify-go-storage-error.md:7]: `Identify go-storage Error` is a not good proposal name to me. How about `Distinguish Errors by IsAosError`?

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**